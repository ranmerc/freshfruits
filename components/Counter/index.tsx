import { AddCircle, RemoveCircle } from "@mui/icons-material";
import { IconButton, Typography } from "@mui/material";
import { MouseEventHandler } from "react";

type CounterProps = {
  count: number;
  incrementLabel: string;
  decrementLabel: string;
  incrementDisabled?: boolean;
  decrementDisabled?: boolean;
  onDecrement: MouseEventHandler<HTMLButtonElement>;
  onIncrement: MouseEventHandler<HTMLButtonElement>;
};

export default function Counter({
  count,
  incrementLabel,
  decrementLabel,
  onDecrement,
  onIncrement,
  incrementDisabled = false,
  decrementDisabled = false,
}: CounterProps) {
  return (
    <>
      <IconButton
        onClick={onDecrement}
        aria-label={decrementLabel}
        disabled={decrementDisabled}
      >
        <RemoveCircle />
      </IconButton>
      <Typography fontSize={"1.2rem"}>{count}</Typography>
      <IconButton
        onClick={onIncrement}
        aria-label={incrementLabel}
        disabled={incrementDisabled}
      >
        <AddCircle />
      </IconButton>
    </>
  );
}
