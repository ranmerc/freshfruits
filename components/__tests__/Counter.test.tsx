import { render, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import Counter from "../Counter";
import renderer, { act } from "react-test-renderer";
import userEvent from "@testing-library/user-event";

describe("Counter component", () => {
  it("Renders with correct details", () => {
    render(
      <Counter
        count={4}
        decrementLabel="increment"
        incrementLabel="decrement"
        onIncrement={jest.fn()}
        onDecrement={jest.fn()}
      />
    );

    expect(screen.getByText("4")).toBeInTheDocument();
    expect(
      screen.getByRole("button", {
        name: "increment",
      })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("button", {
        name: "decrement",
      })
    ).toBeInTheDocument();

    const tree = renderer
      .create(
        <Counter
          count={4}
          decrementLabel="increment"
          incrementLabel="decrement"
          onIncrement={jest.fn()}
          onDecrement={jest.fn()}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("Functions correctly", async () => {
    const increment = jest.fn();
    const decrement = jest.fn();

    render(
      <Counter
        count={4}
        decrementLabel="increment"
        incrementLabel="decrement"
        onIncrement={increment}
        onDecrement={decrement}
      />
    );

    await userEvent.click(screen.getByRole("button", { name: "increment" }));

    waitFor(() => {
      expect(increment).toHaveBeenCalledTimes(1);
    });

    await userEvent.click(screen.getByRole("button", { name: "decrement" }));

    waitFor(() => {
      expect(decrement).toHaveBeenCalledTimes(1);
    });
  });
});
