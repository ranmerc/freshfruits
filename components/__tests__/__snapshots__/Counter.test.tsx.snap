// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component Renders with correct details 1`] = `
[
  <button
    aria-label="increment"
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[MockFunction]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="RemoveCircleIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11H7v-2h10v2z"
      />
    </svg>
  </button>,
  <p
    className="MuiTypography-root MuiTypography-body1 css-b02fjb-MuiTypography-root"
  >
    4
  </p>,
  <button
    aria-label="decrement"
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[MockFunction]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="AddCircleIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
      />
    </svg>
  </button>,
]
`;
